### DISCUSSION TITLE
StandupFormView / EditStandupView preview crashes using bind method for focus

### DISCUSSION THREAD

**Discussion by jmapps2020** [pointfreeco/swift-composable-architecture#1875] [Q&A]
Dear community,

I found that after implementing the bind method from the SwiftUI Navigation Library to bind the focus from the View to the ObservableObject my preview won't work anymore. If I try to use the preview I get a Xcode crash report. 
When I comment out the bind method, the focus does not work but the preview does. Meaning I can click and edit stuff. 
I recently reset my Mac and reinstalled Xcode etc. Nothing works. Can someone help me out?

<img width="1007" alt="Xcode Crash report preview EditStandup : StandupForm" src="https://user-images.githubusercontent.com/69744185/215044103-e52e6f1d-85ea-4395-94e7-ad7571c82fb5.png">

**Comment by mbrandonw** [comment]
Hi @jmapps2020, can you provide a sample project that reproduces the problem so that we can mess with it locally?

Also, if this problem is just with our swiftui-navigation library, and not TCA-related, I move the discussion over to that repo.

Thanks!

**Reply by jmapps2020** [reply]
Hi @mbrandonw, thanks for the quick response. I don't think it's a Library issue. But I just don't know. I get two yellow Xcode warnings in my project. But that is only because I set the concurrency warnings on 'complete' just as Stephen did in the video and has nothing to do with the preview crashing. 

I will provide the sample project in this reply. I hope you can open it. It is just a starting copy of the Standups App where I finished the StandupForm. 
When I comment out the bind method in the StandupFormView I can use the preview. Otherwise it will just crash.

[Standups.xcodeproj.zip](https://github.com/pointfreeco/swift-composable-architecture/files/10521482/Standups.xcodeproj.zip)

If you need me to provide an example project via a repo please let me know as I am not used to providing files and projects this way.

**Reply by jmapps2020** [reply]
Apparently just the .xcodeproj file is not enough. So I will put it on Github and sent you the repo?

**Reply by jmapps2020** [reply]
https://github.com/jmapps2020/Standups. This is the URL. When I provide it in a link it won't work apparently.
Thanks again and I hope you know what the problem might be.

**Reply by mbrandonw** [reply]
Hi @jmapps2020, I'm not able to reproduce any preview problems in the project you included. Can you tell me what steps you take to reproduce?

**Reply by mbrandonw** [reply]
Sorry, by reproduce I mean I can't see the preview crash when I click around the preview and type into the fields. What steps do I have to take in the preview to make it crash?

**Reply by jmapps2020** [reply]
I guess none. When I open the project and navigate to StandupForm and click on the preview screen I get the crash. 
[Schermopname 2023-01-28 om 07.18.35.zip](https://github.com/pointfreeco/swift-composable-architecture/files/10526078/Schermopname.2023-01-28.om.07.18.35.zip)

**Reply by jmapps2020** [reply]
In the video the preview screen turns black. That happens sometimes, otherwise I get the crash report.

**Reply by iampatbrown** [reply]
@jmapps2020 something like this is what @mbrandonw was asking for :)
```
Steps to reproduce:
  1. Wait for preview to load
  2. Tap another attendee <- ⚠️ Crashes here
```
I tested on two laptops as well as different versions of Xcode and things looked like they were working fine... after watching your video I was able to reproduce the crash be tapping on an attendee after the preview had loaded...

Haven't looked into it yet. Just letting people know it's reproducible.

**Reply by jmapps2020** [reply]
Hi @iampatbrown ,

Thank you very much for getting involved and trying to help. I am sorry for my ignorance about 'reproducing'.
So if I am correct you got the preview crash as well without changing the code?

**Reply by iampatbrown** [reply]
@jmapps2020 Yes. The preview crashed without me changing anything. 

It's also reproducible with the following:

```swift
import SwiftUI

class Model: ObservableObject {
  @Published var focus: Field? = .username
  @Published var password: String = ""
  @Published var username: String = "Blob"

  enum Field: Hashable {
    case password
    case username
  }
}

struct ContentView: View {
  @ObservedObject var model: Model
  @FocusState private var focus: Model.Field?

  @State var hasAppeared = false

  var body: some View {
    Form {
      TextField("Username", text: self.$model.username)
        .focused(self.$focus, equals: .username)
      SecureField("Password", text: self.$model.password)
        .focused(self.$focus, equals: .password)
    }
    .onAppear {
      guard !self.hasAppeared else { return }
      self.hasAppeared = true
      guard self.model.focus != self.focus else { return }
      self.focus = self.model.focus
    }
  }
}

struct ContentView_Previews: PreviewProvider {
  static var previews: some View {
    NavigationStack {
      ContentView(model: Model())
    }
  }
}
```

My suspicion is the `NavigationStack` causes an issues with the `FocusState` binding. Possibly causing the view modifier to write to a value that has been invalidated.

**Reply by jmapps2020** [reply]
@iampatbrown , wow that's quick. I have the same. The preview just freezes.

**Reply by iampatbrown** [reply]
@jmapps2020 I'm not sure on the specifics at this stage but it looks related to the fact the `FocusState`'s mirror value can lead to an infinite recursion. SwiftUI is possibly using reflection somewhere which is causing the previews to crash...

edit: Maybe not infinite... a very large reflection that includes the environment... actually maybe there's something different about the preview environment...

**Reply by jmapps2020** [reply]
@iampatbrown 
I really appreciate you getting into it and helping me out. As I am relatively new to programming some terms I don't quite understand yet.😊. For example, what does Mirror Value mean?
For my information, this is a SwiftUI issue which cannot be fixed by us/me?

**Reply by iampatbrown** [reply]
@jmapps2020 Yeah this a SwiftUI bug... I've been trying to find a work around but no luck so far...

Here's a simpler example:
```swift
struct ContentView: View {
  @State var text = "Blob"
  @FocusState var focus

  var body: some View {
    Form {
      TextField("Text", text: $text)
        .focused(self.$focus)
      Button("Tap") {
        self.focus = false
      }
    }
    .onAppear { self.focus = true }
  }
}

struct ContentView_Previews: PreviewProvider {
  static var previews: some View {
    NavigationStack {
      ContentView()
    }
  }
}
```

> what does Mirror Value mean?

A [Mirror](https://developer.apple.com/documentation/swift/mirror) is a representation of a structure as its raw parts. This is often useful for debugging and sometimes lets you see internal properties. For example, we can get a better idea of how `FocusState` works by using `dump(self._focus, maxDepth: 3)`.

```
▿ SwiftUI.FocusState<Swift.Bool>
  - value: false
  ▿ location: Optional(SwiftUI.FocusStoreLocation<Swift.Bool>)
    ▿ some: SwiftUI.FocusStoreLocation<Swift.Bool> #0
      - super: SwiftUI.AnyLocation<Swift.Bool>
      ▹ store: SwiftUI.FocusStore
      ▹ host: Optional(SwiftUI.GraphHost)
      - resetValue: false
      ▹ focusSeed: empty
      - failedAssignment: nil
      ▹ resolvedEntry: Optional(SwiftUI.FocusStore.Entry<Swift.Bool>(value: true, responder: Optional(node(0x0000600002364c00 FocusStateBindingResponder)), bridge: Optional(SwiftUI.FocusBridge), focusScopes: []))
      ▹ resolvedSeed: 37
      - _wasRead: false
  - resetValue: false
```

You can use also [customDump](https://github.com/pointfreeco/swift-custom-dump) as an alternative to `dump` which comes with TCA.

**Reply by jmapps2020** [reply]
@iampatbrown thanks again for explaining and taking the time to do so. I definitely will use this/ practice dump and custom dump in the future and learn more about it.
Indeed the simpler example crashes the preview as well.
I hope you will find a workaround😀.

**Reply by tgrapperon** [reply]
@iampatbrown I'm still unable to reproduce the issue on Xcode 14.2, using any version you extracted. Do you mean it loads, and then crash when you try to interact with the preview?
Using an iPhone 14 pro destination if that matters.

**Reply by iampatbrown** [reply]
@tgrapperon The first example I gave seems to be constantly crashing for me, even after cleaning out derived data and restarting. I've noticed the shorter example sometimes works after cleaning everything and restarting though. I can try reinstall Xcode completely but don't have time for that right now :)

**Reply by iampatbrown** [reply]
@jmapps2020 @tgrapperon Just to confirm, I've tested the first example I posted again on an M1 and Intel machine after clearing all derived data and simulator/preview data. The crash still occurs after trying to change the selection. Either on the first time running the preview or second. 

A current work around for me is to use `.accessibilityHidden(true)` on the preview:
```swift
struct ContentView_Previews: PreviewProvider {
  static var previews: some View {
    NavigationStack {
      ContentView(model: Model())
    }
    .accessibilityHidden(true)
  }
}
```
@jmapps2020 can you see if this works on your end?

**Reply by jmapps2020** [reply]
@iampatbrown unbelievable. The workaround is the solution for me as well.
Is this a common workaround? Really amazing that you found this workaround...
Thank you so much for helping me out. I really appreciate it.

[Schermopname 2023-01-29 om 09.19.05.zip](https://github.com/pointfreeco/swift-composable-architecture/files/10529335/Schermopname.2023-01-29.om.09.19.05.zip)

**Reply by iampatbrown** [reply]
@jmapps2020 I haven't seen or used it as a work around before... I just noticed lots of accessibility calls in the crash report so thought using `.accessibilityHidden(true)` could help.

I might try wipe my laptop and start completely from scratch if no one else is able to reproduce the crash :P It's possible something is different with our machines, maybe a privacy/security/accessibility setting that's causing you and me to have problems and no one else...

**Reply by jmapps2020** [reply]
@iampatbrown that's very clever after reading that report😎.
I did that too. Couple of days ago. Reinstalled Ventura and reset the Mac Hd Disk. 
No luck for me apparently. It is very plausible that our machine settings are slightly different, causing the issue. If you need me to try anything that might work let me know. 
Otherwise good luck trying to get to the bottom of this and I would appreciate any updates😀.

**Reply by jmapps2020** [reply]
@iampatbrown , I have got another one for you. This time it's more complicated I believe. I posted about it on the new Slack channel as well but I also wanted to ask and inform you. So here is the post....

Hi everyone, I have an issue with the preview of the RecordMeeting Feature in the Standups Rebuilt. Running the app in the simulator works just fine. I did reset the previews cache but that didn't help. I hope someone can help me out by looking at the crash report, which I copied into a text editor file.

I am running the latest version of the library. I also get some yellow warnings :warning: saying that .alert has been deprecated, but thats not the issue because in the StandupDetail Feature the alert in the preview works just fine.
To reproduce I added 2 short screen records.

The first one is showing that I am able to skip speakers but when I hit the point of the alert it crashes, as well as I press the end meeting button.

The second one shows that when I set the destination to nil in the endMeetingButtonTapped method, the preview won't crash. Indicating something is wrong with the alert???

I am too much of a beginner to get this wrapped around my head for now thats why I am reaching out for help.

I hope that make things clear enough to move forward.

By the way I had a similar issue with the StandupForm Feature. The workaround there was to put a .accessibilityHidden(true) modifier to the preview. Unfortunately that didn't help here..
Thanks in advance.

[PreviewCrash(1).zip](https://github.com/pointfreeco/swift-composable-architecture/files/10545643/PreviewCrash.1.zip)

[PreviewsCrash(2).zip](https://github.com/pointfreeco/swift-composable-architecture/files/10545645/PreviewsCrash.2.zip)

[Crash Report Preview RecordMeetingPreview.zip](https://github.com/pointfreeco/swift-composable-architecture/files/10545655/Crash.Report.Preview.RecordMeetingPreview.zip)

**Reply by mbrandonw** [reply]
I responded in Slack, but worth mentioning here too.

I've been running into this problem in a completely different project, and it seems related  to presenting a screen twice (whether alert or sheet). For example, to get a crash in RecordMeeting.swift just tap "End meeting", then "Resume", and then "End meeting" again. 

But, if instead you run the StandupDetail.swift preview, start a meeting, and do that same flow (End meeting > Resume > End meeting), it does not crash.

**Reply by jmapps2020** [reply]
Thnx for responding @mbrandonw. Did you get the preview in the other project working at the end? Or is it still an issue...

I didn't try 'deep linking / drilling down' from the StandupDetail into the meeting view but that works indeed. No problems there using the preview. 

In my case, I restart Xcode, open up the project, load the preview from RecordMeetingView, tapping on end meeting and BOOM the crash happens immediately. I have no alert showing whatsoever.

It is so annoying when trying to follow along that these problems occur. I know it is not huge but I just would love to see and test from the previews instead of always running the project in the simulator. 

I haven't filed a report before with Apple. Any guidance on how to do that best? And hopefully a workaround will be found quickly.

**Comment by jmapps2020** [comment]
@mbrandonw,Maybe I am misunderstanding ‘reproduce’ but I just created a repo for the project so I could share it with you via GitHub. If you’re not having any preview issues than it must be a local one on my Mac or Xcode. Problem is that I don’t know where to look for a possible solution..Appreciate your time btw…Op 27 jan. 2023 om 22:41 heeft Brandon Williams ***@***.***> het volgende geschreven:﻿
Hi @jmapps2020, I'm not able to reproduce any preview problems in the project you included. Can you tell me what steps you take to reproduce?

**Reply by tgrapperon** [reply]
Hey @jmapps2020, I'm not able to have an issue neither with the project you've attached. Is there a sequence of actions that makes Xcode crash with this project?
Also, can you post a screenshot of the packages' list with their versions, from the left sidebar in Xcode?

**Reply by jmapps2020** [reply]
<img width="260" alt="Screenshots Xcode warnings" src="https://user-images.githubusercontent.com/69744185/215250569-b92b3ace-bae7-4a67-98e0-b0b829da1b8c.png">
<img width="264" alt="Screenshot Xcode Packages list Standups" src="https://user-images.githubusercontent.com/69744185/215250571-748d4fe7-9ecf-406e-b202-cda7fbe6bbc1.png">

**Reply by jmapps2020** [reply]
[Schermopname 2023-01-28 om 07.18.35.zip](https://github.com/pointfreeco/swift-composable-architecture/files/10526089/Schermopname.2023-01-28.om.07.18.35.zip)

**Reply by jmapps2020** [reply]
Hi @tgrapperon ,

Thanks for helping.

I added screenshots and a recording.
I don't take specific actions.
I just open Xcode, go to the project, navigate to StandupForm and try to interact with the preview

**Reply by tgrapperon** [reply]
@jmapps2020 Thanks for the screenshot. Theses packages versions are the same that I have. It was just to confirm that we were on the same configuration.
